File: homework1/src/warmup.js
// This is homework 1
const rp = require('request-promise');
const crypto = require('crypto');

function change(cents) {
  if (cents < 0) {
    throw new RangeError('amount cannot be negative');
  }
  const result = [];
  let remaining = cents;
  function calc(coin) {
    result.push(Math.floor(remaining / coin));
    remaining %= coin;
  }
  [25, 10, 5, 1].forEach(calc);
  return result;
}

function stripQuotes(str) {
  let final = str.replace(/'/g, '');
  final = final.replace(/"/g, '');
  return final;
}

function scramble(word) {
  if (word.length === 0) {
    return '';
  }
  const newPermutations = Math.floor(Math.random() * word.length);
  const findRandom = word.charAt(newPermutations);
  const newword = (word.slice(0, newPermutations) + (word.slice(newPermutations + 1, word.length)));
  return findRandom + scramble(newword);
}

function powers(base, limit, callback) {
  for (let i = 0; base ** i <= limit; i += 1) {
    const input = base ** i;
    callback(input);
  }
}
function* powersGenerator(base, limit) {
  for (let i = 0; base ** i <= limit; i += 1) {
    const input = base ** i;
    yield input;
}
}
// Thanks @Joey for guidance on this one.
function say(word) {
  if (!word) {
    return '';
  }
  return function phraser(other) {
    if (!other) {
      return word;
    }
    return say(`${word} ${other}`);
  };
}
function interleave(...args){
  let r = args.shift();
  let newr = args;
  let rest = [];
  if (r.length === 0){
    return newr;
  } else if (newr.length ===0) {
    return r;
  } else {
    while (r.length !== 0 && newr.length !==0) {
      rest.push(r.shift());
      rest.push(newr.shift());
      if (r.length === 0) {
        return rest.concat(newr);
      } else if (newr.length === 0) {
      return rest.concat(r);
      }
   }
  }
}
// Thanks to Justin Torres for help with getters
function cylinder(spec) {
  let { radius, height } = spec;
  if (!radius) { radius = 1; }
  if (!height) { height = 1; }
  const surfaceArea = () => (2 * Math.PI * radius * height) + (2 * Math.PI * (radius ** 2));
  const volume = () => (Math.PI * (radius ** 2) * height);
  const stretch = (factor) => { height *= factor; };
  const widen = (factor) => { radius *= factor; };
  const toString = () => `cylinder with radius ${radius} and height ${height}`;

  return Object.freeze({
    volume,
    surfaceArea,
    widen,
    stretch,
    toString,
    get radius() { return radius; },
    get height() { return height; },
  });
}
function makeCryptoFunctions(password, algorithm){
  const localPass = password;
  const localAlg = algorithm;
  return [
    function encrypt(text){

      let cipher = crypto.createCipher(localAlg,localPass);
      let crypt = cipher.update(text,'utf8','hex');
      crypt += cipher.final('hex');
      return crypt;
    },
   
    function decrypt(text){
      let decipher = crypto.createDecipher(localAlg,localPass);
      let decry = decipher.update(text,'hex','utf8');
      decry += decipher.final('utf8');
      return decry;
    }
  ];
}

function randomName(spec) {
  // Get the parameters
  const { gender, region } = spec;
  const options = {
    uri: 'http://uinames.com/api/',
    qs: {
      amount: 1,
      gender,
      region,
    },
    json: true, // Automatically parses the JSON string in the response
  };

  return rp(options)
    .then(({ surname, name }) => `${surname}, ${name}`);
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  makeCryptoFunctions,
  cylinder,
  randomName,
};

File: homework2/warmup.py
# This is homework 2.

import re
import random
import math
from Crypto.Cipher import AES
import requests


def change(cents):
    if cents < 0:
        raise ValueError('amount cannot be negative')
    coins_list = []
    remaining = cents

    def find_coin_number(coin_value, remaining):
        coins_list.append(math.floor(remaining/coin_value))
        return remaining % coin_value
    for value in [25, 10, 5, 1]:
        remaining = find_coin_number(value, remaining)
    return tuple(coins_list)


def strip_quotes(str):
    return re.sub("\"|\'", "", str)


def scramble(words):
    permuted_words = random.sample(words, len(words))
    new_words = ''.join(permuted_words)
    return new_words


def powers(base, max1):
    if (max1 == 0):
        raise StopIteration
    number = 0
    exponent = 0
    while number <= max1:
        yield base ** exponent
        exponent += 1
        number = base ** exponent


def triples(upper_value):
    result = []
    for a in range(1, upper_value+1):
        for b in range(1, upper_value+1):
            for c in range(1, upper_value+1):
                if a ** 2 + b ** 2 == c ** 2 and a < b and b < c:
                    result.append((a, b, c))
    return result


def say(word=None):
    if word is None:
        return ""

    def phrase(new_word=None):
        if new_word is None:
            return word
        return say(word + " " + new_word)
    return phrase


def interleave(first, *second):
    max_num = max(len(first), len(second))
    result = []
    for x in range(0, max_num):
        if (x < len(first)):
            result.append(first[x])
        if (x < len(second)):
            result.append(second[x])
    return result


class Cylinder:
    def __init__(self, height=1, radius=1):
        self.height = height
        self.radius = radius

    @property
    def volume(self):
        return math.pi * (self.radius ** 2) * self.height

    @property
    def surface_area(self):
        return 2 * math.pi * ((self.radius * self.height) + (self.radius ** 2))

    def widen(self, factor):
        self.radius *= factor
        return self

    def stretch(self, factor):
        self.height *= factor
        return self


def make_crypto_functions(key, initialization_vector):
    local_key = key
    local_initialization_vector = initialization_vector

    def encrypt(text):
        cipher = AES.new(local_key, AES.MODE_CBC, local_initialization_vector)
        return cipher.encrypt(text)

    def decrypt(text):
        cipher = AES.new(local_key, AES.MODE_CBC, local_initialization_vector)
        return cipher.decrypt(text)

    return (encrypt, decrypt)


def random_name(gender, region):
    response = requests.get('https://uinames.com/api/',
                            params={'gender': gender,
                                    'region': region,
                                    'amount': 1})
    if 'error' in response.json():
        raise ValueError(response.text)
    return '{}, {}'.format(response.json()['surname'], response.json()['name'])
